go/mysql/conn.go

```diff
@@ -1345,60 +1470,109 @@ func (c *Conn) execQuery(query string, handler Handler, multiStatements bool) (s
 	return remainder, nil
 }
 
-func (c *Conn) execPrepareStatement(stmtID uint32, handler Handler) error {
+// execPrepareStatement runs the query identified by the statement ID, and writes the expected packets to the connection
+// If the client requests that a cursor be opened, we should only write the fields, and wait for subsequent fetch
+// requests to write the rows from the result set.
+func (c *Conn) execPrepareStatement(stmtID uint32, cursorType byte, handler Handler) error {
 	fieldSent := false
-	// sendFinished is set if the response should just be an OK packet.
-	sendFinished := false
+	sendFinished := false // sendFinished is set if the response should just be an OK packet.
 	prepare := c.PrepareData[stmtID]
-	err := handler.ComStmtExecute(c, prepare, func(qr *sqltypes.Result) error {
-		if sendFinished {
-			// Failsafe: Unreachable if server is well-behaved.
-			return io.EOF
-		}
-
-		if !fieldSent {
-			fieldSent = true
+	var err error
+	if cursorType == NoCursor {
+		err = handler.ComStmtExecute(c, prepare, func(qr *sqltypes.Result) error {
+			if sendFinished {
+				// Failsafe: Unreachable if server is well-behaved.
+				return io.EOF
+			}
 
-			if len(qr.Fields) == 0 {
-				sendFinished = true
-				// We should not send any more packets after this.
-				return c.writeOKPacket(qr.RowsAffected, qr.InsertID, c.StatusFlags, 0)
+			if !fieldSent {
+				fieldSent = true
+				if len(qr.Fields) == 0 {
+					sendFinished = true
+					// We should not send any more packets after this.
+					return c.writeOKPacket(qr.RowsAffected, qr.InsertID, c.StatusFlags, 0)
+				}
+				if err = c.writeFields(qr); err != nil {
+					return err
+				}
+				// Now send an EOF packet.
+				if c.Capabilities&CapabilityClientDeprecateEOF == 0 {
+					// With CapabilityClientDeprecateEOF, we do not send this EOF.
+					if err = c.writeEOFPacket(c.StatusFlags, 0); err != nil {
+						return err
+					}
+				}
 			}
-			if err := c.writeFields(qr); err != nil {
-				return err
+			return c.writeBinaryRows(qr)
+		})
+	} else {
+		c.cs = &cursorState{
+			stmtID: stmtID,
+			next: make(chan *sqltypes.Result),
+			done: make(chan error),
+			quit: make(chan error),
+		}
+
+		go func() {
+			err = handler.ComStmtExecute(c, prepare, func(qr *sqltypes.Result) error {
+				// block until query results are sent or receive signal to quit
+				select {
+				case c.cs.next <- qr:
+					return nil
+				case err := <-c.cs.quit:
+					return err
+				}
+			})
+			// pass along error
+			close(c.cs.next)
+			c.cs.done <- err
+		}()
+
+		// Immediately get the very first query result, to write the fields
+		if qr, ok := <- c.cs.next; ok {
+			if !fieldSent {
+				fieldSent = true
+				if len(qr.Fields) == 0 {
+					sendFinished = true
+					err = c.writeOKPacket(qr.RowsAffected, qr.InsertID, c.StatusFlags, 0)
```

Comment:
Should StatusFlag still be CursorExists here?

Reply:
It looks like the CursorExists StatusFlag is unset for MySQL when it the cursor has been exhausted.
So, the server should keep returning that the CursorExists and returns that the cursor doesn't exist when the fetch return no rows. Additionally, it looks like MySQL also returns a [`SERVER_STATUS_LAST_ROW_SENT`](https://mariadb.com/kb/en/ok_packet/#server-status-flag:~:text=SERVER_STATUS_LAST_ROW_SENT); I added this change to vitess as well.